### PR TITLE
Add multi-agent example: tic-tac-toe

### LINKED ISSUE #121: How to support multi-agent reinforcement learning
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website], and in particular read the [known issues]
- [x] I have searched through the [issue tracker] and [issue categories] for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, sys
  print(tianshou.__version__, torch.__version__, sys.version, sys.platform)
  ```

  [source website]: https://github.com/thu-ml/tianshou/
  [known issues]: https://github.com/thu-ml/tianshou/#faq-and-known-issues
  [issue categories]: https://github.com/thu-ml/tianshou/projects/2
  [issue tracker]: https://github.com/thu-ml/tianshou/issues?q=

### PR DESCRIPTION
**Pull request by youkaichao**

### CODE REVIEW

File: examples/multi_agent/MANet.py
Comment (duburcqa) [inline examples/multi_agent/MANet.py]: I would rather explicit `torch.float32` or `torch.float64` since it is a convention specific to each library.

File: examples/multi_agent/MANet.py
Comment (duburcqa) [inline examples/multi_agent/MANet.py]: `unsqueeze(-1)`

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [inline examples/multi_agent/tic_tac_toe.py]: no need of `tolist()` to iterate

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [inline examples/multi_agent/tic_tac_toe.py]: You could directly convert it to `list` since you are no doing any `set` specific operation later.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [inline examples/multi_agent/tic_tac_toe.py]: it should be me necessary for `done` or `reward` to be a `np.ndrray`, but as you wish.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [inline examples/multi_agent/tic_tac_toe.py]: Same. No need to cast to `list` to iterate over rows.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [inline examples/multi_agent/tic_tac_toe.py]: `print(' '.join(map(f, row)))` is better

File: tianshou/env/ma_env.py
Comment (duburcqa) [inline tianshou/env/ma_env.py]: I don't like this class name. Not explicit enough.

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: Same, not explicit enough to my taste.

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: `0` is valid default ?

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: `self.model` should work the same no ?

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: You should be able to `zip` without converting to `list`. Is it more efficient to do so ?

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: Same, is this cast necessary ?

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [inline tianshou/policy/modelfree/MAdqn.py]: No necessary if `legal_actions` is a `np.array` in the first place.

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: The code is copied from other examples. Maybe this can be fixed by another PR.

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: This has to be ``set``, because available actions vary with time. If we hold it with list, they will be converted to ``np.ndarray`` and later be stacked, which will cause an exception (stacking arrays with inequal length).

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: I want to assure the ``set`` only contains python integers. Not sure how a ``set`` of numpy scalar behaves.

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: The code is copied from other examples. Maybe this can be fixed by another PR.

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: I just add this to pass type check.

File: tianshou/policy/modelfree/MAdqn.py
Comment (Trinkle23897) [reply tianshou/policy/modelfree/MAdqn.py]: In the current DQN implementation, 0 is reserved for no target network.

File: tianshou/policy/modelfree/MAdqn.py
Comment (Trinkle23897) [reply tianshou/policy/modelfree/MAdqn.py]: `model` can be "model" or "model_old" (both are string).

File: tianshou/policy/modelfree/MAdqn.py
Comment (youkaichao) [reply tianshou/policy/modelfree/MAdqn.py]: Just to assure they are python objects. It seems okay without ``tolist()``, but this behavior depends on many implicit convertions.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: > I want to assure the set only contains python integers. 

Yes I got that.

> Not sure how a set of numpy scalar behaves.

It should be the same.

File: examples/multi_agent/MANet.py
Comment (duburcqa) [reply examples/multi_agent/MANet.py]: I don't get why to be in a hurry. Why is big deal to do it now ? Since it is not even a required feature on `master` branch.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: Which line are you updating `legal_actions` ?

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [reply tianshou/policy/modelfree/MAdqn.py]: `zip` do not require python object. Only iterable objects. There is no "conversion" involved by default.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: Can you clarify which type check please ?

**Comment by duburcqa** [comment]
@youkaichao As I said, I don't get why to be in a hurry on merge this PR, since your are the only one using it for now.

**Comment by Trinkle23897** [comment]
No, not only him but also other people. Over three people have emailed me or comment on my personal social media to require MARL features, but Kaichao is the first to deal with it.

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: I'm working on this by another pr.

**Comment by duburcqa** [comment]
> No, not only him but also other people. Over three people have emailed me or comment on my personal social media to require MARL features, but Kaichao is the first to deal with it.

So working a bit longer on this feature to improve it shouldn't be an issue if people have been waiting weeks.

**Comment by Trinkle23897** [comment]
> > No, not only him but also other people. Over three people have emailed me or comment on my personal social media to require MARL features, but Kaichao is the first to deal with it.
> 
> So working a bit longer on this feature to improve it shouldn't be an issue if people have been waiting weeks.

Sure.

**Comment by duburcqa** [comment]
If you really want to rush, the good practice is to use a `dev` branch. `master` branch is dedicated to mature features, even when not distributed on pypi and associated with a release version. People wanting to use preview features then should checkout `dev` instead of `master`.

**Comment by duburcqa** [comment]
The good point about `dev` branch, is that there is basically no rules to follow since it only applies on `master`. One can submit and merge PRs for which the CI is even failing, or not having the approval of other members. Yet at the end, when `dev` branch is going to be merged on `master`, it has to follow the rules. It is a good way to release experimental features as fast as possible, without compromising `master` branch.

File: examples/multi_agent/MANet.py
Comment (duburcqa) [reply examples/multi_agent/MANet.py]: Do as you wish, but keep in mind you don't have my approval, even if you don't care about it.

**Comment by Trinkle23897** [comment]
> The good point about `dev` branch, is that there is basically no rules to follow since it only applies on `master`. One can submit and merge PRs for which the CI is even failing, or not having the approval of other members. Yet at the end, when `dev` branch is going to be merged on `master`, it has to follow the rules. It is a good way to release experimental features as fast as possible, without compromising `master` branch.

So I think we can release 0.2.4 first, and then keep `master` branch stable and all the new features will appear in `dev` first. Do you agree?
But I haven't figure out how to deal with API changes, for example, #123: since it will introduce `tianshou.utils.net`, running the newest test with older tianshou core (0.2.3) will raise an ImportError.

**Comment by duburcqa** [comment]
> But I haven't figure out how to deal with API changes, for example, #123: since it will introduce tianshou.utils.net, running the newest test with older tianshou core (0.2.3) will raise an ImportError.

It happens when a library is young. Breaking changes must be avoided if possible, but if it is necessary to improve the quality then it must be done. You can avoid it unfortunately. Please must expect that such things can happen at early stage.

**Comment by duburcqa** [comment]
> So I think we can release 0.2.4 first, and then keep master branch stable and all the new features will appear in dev first. Do you agree?

As you wish. Let's see how it goes after release of 0.2.4.

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: > I don't get why to be in a hurry. Why is big deal to do it now ? Since it is not even a required feature on `master` branch.

I want to keep each PR's integrity by focusing on one theme :)

Comment (youkaichao) [reply examples/multi_agent/MANet.py]: Solved in 185f5dc07bf874d6be63307b6f010f9b8e2d7ad4 . This file will be removed.

Comment (duburcqa) [reply examples/multi_agent/MANet.py]: > I want to keep each PR's integrity by focusing on one theme :)

In my opinion, one theme is not: release something that has flaws easy to fix

But it has already been discussed anyway.

**Review by duburcqa — DISMISSED** [review]

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: ``unsqueeze(-1)`` is not the same as ``view(batch, -1)``.
```
a = torch.ones(4, 5, 6)
b = a.view(a.shape[0], -1)
c = a.unsqueeze(-1)
a.shape
Out[58]: torch.Size([4, 5, 6])
b.shape
Out[59]: torch.Size([4, 30])
c.shape
Out[60]: torch.Size([4, 5, 6, 1])
```

Comment (youkaichao) [reply examples/multi_agent/MANet.py]: I find that ``a.flatten(1)`` does the trick.

**Comment by Trinkle23897** [comment]
Why did you approve this @duburcqa ?

File: examples/multi_agent/MANet.py
Comment (duburcqa) [reply examples/multi_agent/MANet.py]: Indeed I misread and was about to suggest `flatten` :+1:

**Comment by duburcqa** [comment]
> Why did you approve this @duburcqa ?

You said you wanted to merge it as is, no ?

**Comment by Trinkle23897** [comment]
I'm not willing to merge now because I think the interface of MA is somewhat hardcode. In my mind, the normal MAPolicy will accept a list of policies and a dispatch function to indicate how to split the full data into N agent's view.
But this brings some other issues: if we use PPO+DQN as two agents, since PPO is an on-policy algo and DQN is off-policy, how can we deal with their sample function with regard to a single buffer? It can be a future work.

**Comment by duburcqa** [comment]
> But this brings some other issues: if we use PPO+DQN as two agents, since PPO is an on-policy algo and DQN is off-policy, how can we deal with their sample function with regard to a single buffer? It can be a future work.

That's not an easy task and probably it cannot be solved rapidly. Did you have a look to how the libraries supporting this feature were handling it ?

**Comment by Trinkle23897** [comment]
> > But this brings some other issues: if we use PPO+DQN as two agents, since PPO is an on-policy algo and DQN is off-policy, how can we deal with their sample function with regard to a single buffer? It can be a future work.
> 
> That's not an easy task and probably it cannot be solved rapidly. Did you have a look to how the libraries supporting this feature were handling it ?

Agree. I'll have a second check with this pr.

File: examples/multi_agent/MANet.py
Comment (youkaichao) [reply examples/multi_agent/MANet.py]: Fixed in ba7c184b83bbdc9ac960aaab99b9835f78d1c42e . This file will be removed.

**Comment by Trinkle23897** [comment]
@youkaichao please add the corresponding documentation in https://tianshou.readthedocs.io/en/latest/tutorials/cheatsheet.html, also with minor modification in README.md.

**Comment by duburcqa** [comment]
@Trinkle23897 Can you enable "Dismiss stale pull request approvals when new commits are pushed " on master please ?

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: They are not updated. They becomes a field in observation. In fact, you can try to use a ``list`` and run the script, which will lead to an exception.

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: The function signature of ``step()``.

File: tianshou/policy/modelfree/MAdqn.py
Comment (youkaichao) [reply tianshou/policy/modelfree/MAdqn.py]: ``legal_actions `` here is a ``set``.  In the below, ``random.choice`` requires an object that supports indexing.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: I don't tell you to use a `list` of course, but to use 
```
legal_actions = list({i for i, b in enumerate(
            self.current_board.flatten().tolist()) if b == 0})
```
since you are not relying on `set` properties elsewhere, and it would remove several conversion that comes later.

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: Fixed in 279fd28 .

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: OK I see. This is problematic, since it goes against the specification of OpenAI Gym:
https://github.com/openai/gym/blob/345c65973fc7160d8be374745a60c36869d8accc/gym/core.py#L52

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: Fixed in 279fd28 .

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: Fixed in 279fd28 .

File: tianshou/policy/modelfree/MAdqn.py
Comment (duburcqa) [reply tianshou/policy/modelfree/MAdqn.py]: Yes of course, but I mean it could be a list in the first place I think (see my comment above).

File: tianshou/env/ma_env.py
Comment (youkaichao) [reply tianshou/env/ma_env.py]: Fixed in 279fd28 .

File: tianshou/policy/modelfree/MAdqn.py
Comment (youkaichao) [reply tianshou/policy/modelfree/MAdqn.py]: Fixed in 279fd28 .

File: tianshou/policy/modelfree/MAdqn.py
Comment (youkaichao) [reply tianshou/policy/modelfree/MAdqn.py]: Fixed in 279fd28 .

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: > I don't tell you to use a `list` of course, but to use
> 
> ```
> legal_actions = list({i for i, b in enumerate(
>             self.current_board.flatten().tolist()) if b == 0})
> ```
> 
> since you are not relying on `set` properties elsewhere, and it would remove several conversion that comes later.

I tested your code, it raises an exception. You can try it, too :)

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: > OK I see. This is problematic, since it goes against the specification of OpenAI Gym:
> https://github.com/openai/gym/blob/345c65973fc7160d8be374745a60c36869d8accc/gym/core.py#L52

I do not think this is problematic. In the future, we may support multi-agent environment whose reward is a vector (with each item the reward of each player).

**Comment by youkaichao** [comment]
> @youkaichao please add the corresponding documentation in https://tianshou.readthedocs.io/en/latest/tutorials/cheatsheet.html, also with minor modification in README.md.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: I don't have time to check by myself, I trust you on that.

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: It is, in a sense that you require something that it is not part of OpenAI Gym, so Tianshou RL algorithm will raise exception on any environment implementing the specification of OpenAI Gym in one day typing becomes strict and (almost) not only cosmetic as it is the case for now.

Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: > It is, in a sense that you require something that it is not part of OpenAI Gym, so Tianshou RL algorithm will raise exception on any environment implementing the specification of OpenAI Gym in one day typing becomes strict and (almost) not only cosmetic as it is the case for now.

So you mean we must stay in the scope of OpenAI Gym? That seems too restricted, as multi-agent RL is apparently not in the scope of OpenAI Gym.

File: examples/multi_agent/tic_tac_toe.py
Comment (youkaichao) [reply examples/multi_agent/tic_tac_toe.py]: > I don't have time to check by myself, I trust you on that.

You do not need to trust me. The code fails instantly, so you can try yourself and trust yourself :)

It seems strange: I think the conclusion is clear by checking the concept of Batch, but apparently it is not the case for you. Does it mean that we have some misunderstanding on the concept of ``Batch``?

**Comment by Trinkle23897** [comment]
> @Trinkle23897 Can you enable "Dismiss stale pull request approvals when new commits are pushed " on master please ?

Done

File: examples/multi_agent/tic_tac_toe.py
Comment (duburcqa) [reply examples/multi_agent/tic_tac_toe.py]: > > It is, in a sense that you require something that it is not part of OpenAI Gym, so Tianshou RL algorithm will raise exception on any environment implementing the specification of OpenAI Gym in one day typing becomes strict and (almost) not only cosmetic as it is the case for now.
> 
> So you mean we must stay in the scope of OpenAI Gym? That seems too restricted, as multi-agent RL is apparently not in the scope of OpenAI Gym.

Checking this needs to go throughout all the code, and I do not have that much time. An issue #127 is recording this. I shall move on to finish the marl example.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: THis is nice, but are you sure it is the right PR to add this ?

Comment (duburcqa) [reply tianshou/data/batch.py]: Also, I would clarify the name and call it `_assert_type_keys`, since it is actually not checking keys but asserting keys types.

File: tianshou/env/ma_env.py
Comment (duburcqa) [inline tianshou/env/ma_env.py]: typo.

File: tianshou/policy/multiagent/madqn.py
Comment (duburcqa) [inline tianshou/policy/multiagent/madqn.py]: skiping line after `self` is better in my opinion.

**Comment by duburcqa** [comment]
> > @Trinkle23897 Can you enable "Dismiss stale pull request approvals when new commits are pushed " on master please ?
>
> Done

Thank you !

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Please check pull reques #126 .

**Comment by youkaichao** [comment]
This PR contains some code from #130 and #125 , so it should be reviewed after the merge of those two prs.

**Comment by youkaichao** [comment]
This is a draft version, no need to review now.

**Comment by duburcqa** [comment]
> This is a draft version, no need to review now.

Yes, it is usually what I expect from a WIP pull request :smile: (beside, you should sleep a bit :stuck_out_tongue_closed_eyes: )

File: tianshou/policy/modelfree/dqn.py
Comment (Trinkle23897) [inline tianshou/policy/modelfree/dqn.py]: rew_norm=False

File: test/multiagent/test_tic_tac_toe.py
Comment (youkaichao) [reply test/multiagent/test_tic_tac_toe.py:7]: In my opinion, for multi-agent RL, ``lambda x: x[args.agent_id - 1]`` can be called "default". So I think it is okay.

**Comment by youkaichao** [comment]
This PR now enables Tianshou of multi-agent RL with conditional moves and cyclic moves. For simultaneous moves (moba games), there are still some unsettled issues, like #136 .

**Comment by youkaichao** [comment]
> This PR contains some code from #130 and #125 , so it should be reviewed after the merge of those two prs.

Ready for merge now.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: No need to change. Or at least improve the exception handling: don't use generic `Exception` nor raw `print`.

File: tianshou/data/batch.py
Comment (duburcqa) [inline tianshou/data/batch.py]: It should be useless now.

File: tianshou/policy/multiagent/mapolicy.py
Comment (duburcqa) [inline tianshou/policy/multiagent/mapolicy.py]: Replace e by the actual content:
```
holder = Batch.cat([{'act': act} for (has_data, agent_index, out, act, each_state) in results if has_data])
```

It is longer but much clearer.

File: tianshou/policy/multiagent/mapolicy.py
Comment (duburcqa) [inline tianshou/policy/multiagent/mapolicy.py]: Remove `condense`

File: tianshou/policy/multiagent/mapolicy.py
Comment (youkaichao) [reply tianshou/policy/multiagent/mapolicy.py]: Fixed in 762c57e .

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Sorry, this is my debug code. removed in 8cf0d13 .

File: tianshou/data/batch.py
Comment (youkaichao) [reply tianshou/data/batch.py]: Removed in 8cf0d13 .

File: tianshou/policy/multiagent/mapolicy.py
Comment (youkaichao) [reply tianshou/policy/multiagent/mapolicy.py]: Removed in 8cf0d13 .

**Review by duburcqa — DISMISSED** [review]

**Comment by Trinkle23897** [comment]
I'll review it in the next two hours.
Many docs need to be fixed.

**Comment by Trinkle23897** [comment]
3 things to do:
- ~~Batch.stack/cat with torch.cuda.Tensor~~ the bug still exists but I bypass it.
- ~~remove MADQN, add mask to DQN~~ done
- ~~modify tutorial~~ done

File: tianshou/policy/modelfree/dqn.py
Comment (youkaichao) [inline tianshou/policy/modelfree/dqn.py]: type: ww

File: tianshou/policy/modelfree/dqn.py
Comment (youkaichao) [inline tianshou/policy/modelfree/dqn.py]: ``obs_ = obs.obs if hasattr(obs, 'obs') else obs`` is more clear.

File: tianshou/policy/modelfree/dqn.py
Comment (youkaichao) [inline tianshou/policy/modelfree/dqn.py]: maybe it is better to use a flag. ``if hasattr(obs, 'mask'):`` occurs several times.

File: tianshou/policy/modelfree/dqn.py
Comment (youkaichao) [inline tianshou/policy/modelfree/dqn.py]: define the mask to be of np.bool_ type, then just ``q_[mask]`` is enough.

File: tianshou/policy/modelfree/dqn.py
Comment (youkaichao) [inline tianshou/policy/modelfree/dqn.py]: nice conversion!

File: tianshou/policy/multiagent/mapolicy.py
Comment (youkaichao) [inline tianshou/policy/multiagent/mapolicy.py]: set to ``0`` is not safe. set to ``-np.inf`` instead. Mask is a strong constraint. However, if there are zeros in logits, this function may return illegal actions.

**Review by Trinkle23897 — APPROVED** [review]

**Comment by Trinkle23897** [comment]
@duburcqa should be okay now, please have a check.

**Review by duburcqa — APPROVED** [review]

**Pull request merged**
